src/test/tst_attrsgenerator.cpp

```diff
@@ -31,61 +31,434 @@ class TestAttrsGenerator: public QObject
 private slots:
     void initTestCase() {}
     void cleanupTestCase() {}
-    void tst_parseStarCmd();
-    void tst_parseHashCmd();
+    void tst_parseStarCmd_min();
+    void tst_parseStarCmd_max();
+    void tst_parseStarCmd_rand();
+    void tst_parseHashCmd_min();
+    void tst_parseHashCmd_max();
+    void tst_parseHashCmd_rand();
+    void tst_parseHashCmd_setValue();
+    void tst_parseHashCmd_mixedFunc();
+    void _tst_attrs(SetOfAttributes res, Attributes attrs,  bool testValues);
+    void _tst_mode(SetOfAttributes res, QString mode, AttributeRange* attrRge[], int numOfAttrRges);
 };
 
-void TestAttrsGenerator::tst_parseStarCmd()
+void TestAttrsGenerator::_tst_attrs(SetOfAttributes res, Attributes attrs, bool testValues)
 {
-    QString error;
+    for(int i = 0; i < res.size(); i++){
+        QCOMPARE(res.at(i).names(), attrs.names());
+        QCOMPARE(res.at(i).size(), attrs.size());
 
-    // - same mode for all attributes:
-    //         '*integer;[min|max|rand_seed]'
-    /*
-    AttributesScope attrsScope =
-    QString cmd =
-    AttrsGenerator* agen = AttrsGenerator::parse(attrsScope,cmd, error);
-    */
-    /** IMPORTANT: this function is likely to be very long,
-     * if so, consider breaking into smaller functions
-     * (perhaps tst_parseStarCmd_min(), tst_parseStarCmd_max() ...) **/
+        if(testValues){
+            QCOMPARE(res.at(i).values(), attrs.values());
+            for(int j = 0; j < attrs.size(); j++){
+                QCOMPARE(res.at(i).value(j), attrs.value(j));
+            }
+        }
+    }
+}
+
+void TestAttrsGenerator::_tst_mode(SetOfAttributes res, QString mode, AttributeRange* attrRge[], int numOfAttrRges)
```

Comment:
You could pass the `AttrsScope ascope` instead of both `AttributeRange* attrRge[]` and `int numOfAttrRges`. Also, as we don't change any of them, it's better to pass them by a const reference. In this way we don't make unnecessary copies of their values.
``` c++
void TestAttrsGenerator::_tst_mode(const SetOfAttributes& res, const QString& mode, const AttrsScope& ascope)
```

The loop would them looks like:
``` c++
if (mode == "min") {
    for (const auto& attrs : setOfAttrs) {
        for (int i = 0; i < attrs.size(); ++i) {
            auto attrRge = ascope.value(attrs.name(i));
            QCOMPARE(attrs.value(i), ar->min());
        }
    }
}...
```